ethexe/observer/src/observer.rs

```diff
@@ -267,107 +204,91 @@ pub(crate) async fn read_block_events_batch(
-                mirror::events::signatures::ALL
-                    .iter()
-                    .map(|hash| B256::new(hash.to_fixed_bytes())),
-            ));
-
-        let mirrors_logs_fut = provider.get_logs(&mirrors_events_filter);
-
-        let (router_logs, wvara_logs, mirrors_logs) =
-            future::join3(router_logs_fut, wvara_logs_fut, mirrors_logs_fut).await;
-        let (router_logs, wvara_logs, mirrors_logs) = (router_logs?, wvara_logs?, mirrors_logs?);
-
-        for router_log in router_logs {
-            let block_hash = router_log
-                .block_hash
-                .ok_or(anyhow!("Block hash is missing"))?
-                .0
-                .into();
-
-            let Some(router_event) = router::events::try_extract_event(router_log)? else {
-                continue;
-            };
-
-            events_map
-                .entry(block_hash)
-                .or_default()
-                .push(router_event.into());
-        }
+        let end_block = to_block.min(start_block + MAX_QUERY_BLOCK_RANGE - 1);
 
-        for wvara_log in wvara_logs {
-            let block_hash = wvara_log
-                .block_hash
-                .ok_or(anyhow!("Block hash is missing"))?
-                .0
-                .into();
+        let filter = Filter::new().from_block(start_block).to_block(end_block);
 
-            let Some(wvara_event) = wvara::events::try_extract_event(wvara_log)? else {
-                continue;
-            };
+        let iter_res = read_events_impl(router_address, wvara_address, provider, filter).await?;
 
-            events_map
-                .entry(block_hash)
-                .or_default()
-                .push(wvara_event.into());
+        res.extend(iter_res.into_iter());
+
+        start_block = end_block + 1;
+    }
+
+    Ok(res)
+}
+
+async fn read_events_impl(
+    router_address: AlloyAddress,
+    wvara_address: AlloyAddress,
+    provider: &ObserverProvider,
+    filter: Filter,
+) -> Result<HashMap<H256, Vec<BlockEvent>>> {
+    let router_and_wvara_topic = Topic::from_iter(
+        router::events::signatures::ALL
+            .into_iter()
+            .chain(wvara::events::signatures::ALL)
+            .map(|v| v.to_fixed_bytes().into()),
+    );
+
+    let router_and_wvara_filter = filter
+        .clone()
+        .address(vec![router_address, wvara_address])
+        .event_signature(router_and_wvara_topic);
+
+    let mirror_filter = filter.event_signature(Topic::from_iter(
+        mirror::events::signatures::ALL.map(|v| v.to_fixed_bytes().into()),
+    ));
+
+    let (router_and_wvara_logs, mirrors_logs) = future::try_join(
+        provider.get_logs(&router_and_wvara_filter),
+        provider.get_logs(&mirror_filter),
+    )
+    .await?;
+
+    let block_hash_of = |log: &alloy::rpc::types::Log| -> Result<H256> {
+        log.block_hash
+            .map(|v| v.0.into())
+            .ok_or(anyhow!("Block hash is missing"))
```

Comment:
mb use ok or else here?

Reply:
seems that we use value creation everywhere in the code, moreover its still super cheap